File: client/src/geo.ts
export type LonLat = [lon: number, lat: number];
export type LatLon = [lat: number, lon: number];

export interface LatLonLocation {
  lat: number;
  lon: number;
}

export function haversineMeters(from: LonLat, to: LonLat): number {
  const r = 6_371_000;
  const [lon1, lat1] = from.map((v) => v * Math.PI / 180) as LonLat;
  const [lon2, lat2] = to.map((v) => v * Math.PI / 180) as LonLat;
  const dLat = lat2 - lat1;
  const dLon = lon2 - lon1;
  const a = Math.sin(dLat / 2) ** 2
    + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return 2 * r * Math.asin(Math.sqrt(a));
}

export function distanceMeters(a: LatLonLocation, b: LatLonLocation): number {
  return haversineMeters([a.lon, a.lat], [b.lon, b.lat]);
}

File: server/src/geo.rs
pub fn haversine_m(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let r = 6_371_000.0;
    let dlat = (lat2 - lat1).to_radians();
    let dlon = (lon2 - lon1).to_radians();
    let lat1 = lat1.to_radians();
    let lat2 = lat2.to_radians();
    let a = (dlat / 2.0).sin().powi(2) + lat1.cos() * lat2.cos() * (dlon / 2.0).sin().powi(2);
    2.0 * r * a.sqrt().asin()
}

#[cfg(test)]
mod tests {
    use super::haversine_m;

    #[test]
    fn haversine_is_zero_for_same_point() {
        assert_eq!(haversine_m(47.6062, -122.3321, 47.6062, -122.3321), 0.0);
    }

    #[test]
    fn haversine_estimates_known_city_distance() {
        let seattle_to_portland_m = haversine_m(47.6062, -122.3321, 45.5152, -122.6784);
        assert!((seattle_to_portland_m - 233_000.0).abs() < 2_000.0);
    }
}
